src/modules/cb_inbound_messaging.erl

```
@@ -0,0 +1,328 @@
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc
+%% @end
+%%------------------------------------------------------------------------------
+-spec on_successful_validation(cb_context:context(), path_token()) -> cb_context:context().
+on_successful_validation(Context, Carrier) ->
+    Setters = [{fun carrier_validate/2, Carrier}
+              ,fun to_number_exists/1
+              ,fun determine_account/1
+              ,fun account_is_enabled/1
+              ,fun account_is_in_good_standing/1
+              ,fun account_has_messaging_enabled/1
+              ,fun reseller_is_in_good_standing/1
+              ,fun reseller_has_messaging_enabled/1
+              ,fun create_request/1
+              ],
+    cb_context:validators(Context, Setters).
+
+%%------------------------------------------------------------------------------
+%% validation callbacks
+%%
+%%------------------------------------------------------------------------------
+
+-spec carrier_validate(cb_context:context(), path_token()) -> cb_context:context().
+carrier_validate(Context, Carrier) ->
+    KzdCarrier = binary_to_atom(<<"knm_", Carrier/binary>>),
+    Payload = cb_context:req_data(Context),
+    case kz_module:is_exported(KzdCarrier, 'validate_im', 1) of
+        true ->
+            handle_carrier_validate(Context
+                                   ,KzdCarrier:validate_im(Payload)
+                                   );
+        _ ->
+            cb_context:add_system_error('bad_identifier', Carrier, Context)
+    end.
+
+handle_carrier_validate(Context, {error, _Error}) ->
+    cb_context:add_system_error('parse_error', Context);
+handle_carrier_validate(Context, {ok, JObj}) ->
+    cb_context:set_doc(Context, JObj).
+
+-spec to_number_exists(cb_context:context()) -> cb_context:context().
+to_number_exists(Context) ->
+    Doc = cb_context:doc(Context),
+    [Number] = kz_im:to(Doc),
```

What enforces 1 and only 1 element in the list?

---

That's a good point.
TIO only ever sends one item in the list, if there are multiple recipients it uses the 'other recipients' field. TIO sends separate hooks for each recipient though. 

If they want to change that it'll be a major code change and should fail until fixed, I'll enforce this as part of validation.

---

Not following why it is a major change? here you only need to loop over the list, no?

---

It'd be a larger change on their end than ours, but we have to reflect it in a fair number of ways.
Biggest problem I see is if not all the `to` numbers exist in kazoo or if they can't be routed for other reasons. How do we want to handle a partial success?

---

Further issues:
if one number is invalid, `cb_sms` just fails the whole request. It's not that simple here.
cb_sms has that luxury because it's expected all the numbers are from the same account. If one number fails because there's insufficient funds, all of them do. Fail the request.
With tio a single valid hook could span multiple end-user accounts, each with different funding statuses. If TIO sends a list of `to` numbers for one reseller across five end-user accounts, two of which have sms disabled, one which has insufficient funds, and two which are in good standing and can do sms, how should we respond?
we shouldn't fail the whole request, two 'innocent' accounts wouldn't get messages because a sibling account isn't in good standing. At a minimum we'd need to turn the whole validation structure upside down to deal with that.

This also makes the 'different reseller's numbers' thing far more complicated on TIO's side. But that's their problem.

I could see a world where we do get to fail the whole request, but we'd still need to compose a response that lets tio remove offending numbers and try again, and that also requires turning the validation structure upside down to deal with it.

So skip it until someone wants it.

---

my opinion is to fail the request with the failing numbers, as you suggest. If TIO wants to offer a request with bogus data, regardless of valid data that may also be contained therein, KAZOO should reject the request. We don't accept a request with schema validation errors because the other fields are passing fine.

---

the code is updated now, now it tracks the account id of the first number, if it changes will fail with proper validation error.

As a reminder, the initial account id on the URL is expected to be the reseller of the number. This is how tio webhook is created by kazoo-trunkingio app.

This endpoint should lookup the account of the first number and make sure the reseller is is in hierarchy then check of all numbers are belong to the same account. multiple number across different account is not supported.

Also we checked with tio, they indeed send it a single number array, even if it is a group message, it fires one webhook per number. This code was causing issues previously, but since it is new code, and we just try to sanity check stuff.